x/delayedack/keeper/invariants.go

```diff
@@ -0,0 +1,108 @@
+)
+
+// RegisterInvariants registers the delayedack module invariants
+func RegisterInvariants(ir sdk.InvariantRegistry, k Keeper) {
+	ir.RegisterRoute(types.ModuleName, "rollapp-finalized-packet", PacketsFromFinalizedHeightsAreFinalized(k))
+	ir.RegisterRoute(types.ModuleName, "rollapp-reverted-packet", PacketsFromRevertedHeightsAreReverted(k))
+}
+
+// AllInvariants runs all invariants of the x/delayedack module.
+func AllInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		res, stop := PacketsFromFinalizedHeightsAreFinalized(k)(ctx)
+		if stop {
+			return res, stop
+		}
+		res, stop = PacketsFromRevertedHeightsAreReverted(k)(ctx)
+		if stop {
+			return res, stop
+		}
+		return "", false
+	}
+}
+
+// PacketsFromFinalizedHeightsAreFinalized checks that all rollapp packets stored for a rollapp finalized height are also finalized
+func PacketsFromFinalizedHeightsAreFinalized(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var msg string
+		rollapps := k.rollappKeeper.GetAllRollapps(ctx)
+
+		rollappsFinalizedHeight := make(map[string]uint64)
+		for _, rollapp := range rollapps {
+			latestFinalizedStateIndex, found := k.rollappKeeper.GetLatestFinalizedStateIndex(ctx, rollapp.RollappId)
+			if !found {
+				continue
+			}
+			latestFinalizedStateInfo, found := k.rollappKeeper.GetStateInfo(ctx, rollapp.RollappId, latestFinalizedStateIndex.Index)
+			if !found {
+				continue
+			}
+			rollappsFinalizedHeight[rollapp.RollappId] = types.GetLastestFinalizedHeigtFromStateInfo(&latestFinalizedStateInfo)
+
+		}
+
+		packets := k.GetAllRollappPackets(ctx)
+
+		for _, packet := range packets {
+			latestFinalizedHeight := rollappsFinalizedHeight[packet.RollappId]
+			if latestFinalizedHeight == 0 {
+				continue
+			}
+			if packet.ProofHeight <= latestFinalizedHeight && packet.Status != commontypes.Status_FINALIZED {
+				msg += fmt.Sprintf("rollapp packet for the height should be in finalized status. height=%d, rollapp=%s, status=%s\n", packet.ProofHeight, packet.RollappId, packet.Status)
+				return msg, true
+			}
+		}
+		return msg, false
+	}
+}
+
+// PacketsFromRevertedHeightsAreReverted checks that all rollapp packets stored for a rollapp reverted height are also reverted
+func PacketsFromRevertedHeightsAreReverted(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var msg string
+
+		rollapps := k.rollappKeeper.GetAllRollapps(ctx)
+
+		for _, rollapp := range rollapps {
+			latestFinalizedStateIndex, found := k.rollappKeeper.GetLatestFinalizedStateIndex(ctx, rollapp.RollappId)
+			if !found {
+				continue
+			}
+
+			stateInfoIndex := latestFinalizedStateIndex.Index + 1
+			// Checking that all packets after the latest finalized height, that belong to a reverted state info, are also in reverted state
+			for {
+				stateInfoToCheck, found := k.rollappKeeper.GetStateInfo(ctx, rollapp.RollappId, stateInfoIndex)
+				stateInfoIndex++
```

Comment:
why u checking all the states of the rollapp? if it's reverted, all the states after the latest finalized will be reverted. it's rollapp module invariant.

I think u can check rollapp.Frozen for reverted rollapp instead of checking the state info

Reply:
updated